reference/security/encryption-at-rest.md

```diff
@@ -0,0 +1,126 @@
+
+Encryption-at-rest means that data is encrypted when it is stored. As opposed to encryption in flight (TLS) or encryption in use (rarely used). Different things could be doing encryption-at-rest (SSD drive, file system, cloud vendor, etc), but if the database does encryption by itself this helps ensure that attackers must authenticate with the database to gain access to data. For example, even when an attacker gains access to the physical machine, he cannot access data by copying files on disk.
+
+TiKV supports encryption-at-rest starting from v4.0.0. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption-at-rest, an encryption key must be provided by user and this key is called master key. The suggested way to provide the master key is via AWS KMS, but specifying a key in plaintext stored in a file is also supported. TiKV automatically rotate data keys that it used to encrypt actual data files. Manually rotating the master key can be done occassionally. Note that encryption-at-rest only encrypts data at rest (i.e. on disk) and not while data is transferred over network. It is advices to use TLS together with encryption-at-rest.
+
+Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backup to S3. A customer owned AWS KMS key can also be used together with S3 server-side encrytion.
+
+## TiKV Encryption-At-Rest
+
+### Overview
+
+TiKV currently support encrypting data using AES128, AES192 or AES256, in CTR mode. There are two types of keys used in TiKV when encryption is enabled:
+
+1. Master key. The master key is provided by user and is used to encrypt the data keys TiKV generates. Management of master key is external to TiKV.
+2. Data key. The data key is generated by TiKV and is the key actually being used to encrypt data. The data key is automatically rotated by TiKV.
+
+The same master key can be shared by multiple instances of TiKV. The recommended way to provide a master key in production is via AWS KMS. User would need to create a CMK through AWS KMS, and then provide the CMK key id to TiKV in config file. The TiKV process would need access to the KMS CMK while it is running, which can be done by using [IAM](https://aws.amazon.com/iam/). If TiKV fail to get access to the KMS CMK, it will failed to start or restart. If TiKV loss access to the KMS CMK while it is running, data key rotation will be temporarily disabled. Please refer to AWS documentation for KMS and IAM usage.
+
+Alternatively, if using custom key is desired, supplying master key via file is also supported. The file need to contain a 256 bits (or 32 bytes) key encoded as hex string. The file should end with a newline (i.e. "\n") and contain nothing else. Persistenting the key on disk, however, leaks the key, so the key file is only suitable to be stored on tempfs.
```

Comment:
```suggestion
Alternatively, if using custom key is desired, supplying the master key via file is also supported. The file needs to contain a 256 bits (or 32 bytes) key encoded as hex string. The file should end with a newline (i.e. "\n") and contain nothing else. Persisting the key on disk, however, leaks the key, so the key file is only suitable to be stored on tempfs in RAM.
```

Reply:
Why do we require a newline? It seems like we could chomp new lines if they exist?

Reply:
I would like to make the file format very strict to make sure user aware of the required key length. I would like to do without the newline, but text editors (at least vim) tend to add a extra newline at line end, which is not a bad idea to indicate end of the key.

Reply:
If I had to choose one, I would probably not have a newline. I generally agree that being strict about inputs is a much better approach, but chomping newlines is the one thin I don't see a problem with. I don't think it will cause any confusion about the file format and they can potentially avoid some deployment delays.

Reply:
we can make the newline optional, but it is too late for 4.0 GA branch cut.

Reply:
file key are not for production use anyway.